diff --git a/plugins/housekeeping/gsd-disk-space.c b/plugins/housekeeping/gsd-disk-space.c
index 9d57531..8a00aaf 100644
--- a/plugins/housekeeping/gsd-disk-space.c
+++ b/plugins/housekeeping/gsd-disk-space.c
@@ -1050,7 +1050,7 @@ gsd_ldsm_setup (gboolean check_now)
                                                  ldsm_check_all_mounts, NULL);
         g_source_set_name_by_id (ldsm_timeout_id, "[gnome-settings-daemon] ldsm_check_all_mounts");
 
-        purge_trash_id = g_timeout_add_seconds (3600, ldsm_purge_trash_and_temp, NULL);
+        purge_trash_id = g_timeout_add_seconds (60, ldsm_purge_trash_and_temp, NULL);
         g_source_set_name_by_id (purge_trash_id, "[gnome-settings-daemon] ldsm_purge_trash_and_temp");
 }
 
