diff --git a/plugins/power/gsd-power-manager.c b/plugins/power/gsd-power-manager.c
index c93c620..90a2f01 100644
--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -2038,6 +2038,38 @@ gsd_power_manager_class_init (GsdPowerManagerClass *klass)
         notify_init ("gnome-settings-daemon");
 }
 
+static void
+wait_for_running_screensaver (GsdPowerManager *manager)
+{
+        gint idx = 0;
+        GVariant *res;
+        GError *error = NULL;
+        guint active_time = 0;
+
+        while (idx < 30) {
+                error = NULL;
+                res = g_dbus_proxy_call_sync (G_DBUS_PROXY (manager->screensaver_proxy),
+                                                      "GetActiveTime", NULL,
+                                                      G_DBUS_CALL_FLAGS_NONE,
+                                                      -1, NULL, &error);
+
+                if (error) {
+                        g_warning ("Could not ask screensaver to activate time: %s", error->message);
+                        g_error_free (error);
+                        break;
+                }
+
+                g_variant_get (res, "(u)", &active_time);
+                g_variant_unref (res);
+
+                if (active_time > 0)
+                        break;
+
+                g_usleep (100 * 1000);
+                idx++;
+        }
+}
+
 static void
 handle_screensaver_active (GsdPowerManager *manager,
                            GVariant        *parameters)
@@ -2054,8 +2086,10 @@ handle_screensaver_active (GsdPowerManager *manager,
                  * and its fade has finished.
                  *
                  * See also idle_configure() */
-                if (active)
+                if (active) {
+                        wait_for_running_screensaver (manager);
                         idle_set_mode (manager, GSD_POWER_IDLE_MODE_BLANK);
+                }
         }
 }
 
@@ -2493,7 +2527,9 @@ on_randr_event (GnomeRRScreen *screen, gpointer user_data)
 static void
 handle_suspend_actions (GsdPowerManager *manager)
 {
-        backlight_disable (manager);
+        wait_for_running_screensaver (manager);
+
+//        backlight_disable (manager);
         uninhibit_suspend (manager);
 }
 
@@ -2501,7 +2537,7 @@ static void
 handle_resume_actions (GsdPowerManager *manager)
 {
         /* ensure we turn the panel back on after resume */
-        backlight_enable (manager);
+//        backlight_enable (manager);
 
         /* set up the delay again */
         inhibit_suspend (manager);
