diff --git a/plugins/housekeeping/gsd-disk-space.c b/plugins/housekeeping/gsd-disk-space.c
index 8a00aaf..863009a 100644
--- a/plugins/housekeeping/gsd-disk-space.c
+++ b/plugins/housekeeping/gsd-disk-space.c
@@ -1038,6 +1038,8 @@ gsd_ldsm_setup (gboolean check_now)
         gsd_ldsm_get_config ();
         g_signal_connect (G_OBJECT (settings), "changed",
                           G_CALLBACK (gsd_ldsm_update_config), NULL);
+        g_signal_connect (G_OBJECT (privacy_settings), "changed",
+                          G_CALLBACK (gsd_ldsm_update_config), NULL);
 
         ldsm_monitor = g_unix_mount_monitor_get ();
         g_signal_connect (ldsm_monitor, "mounts-changed",
